dialogue-serde/src/main/java/com/palantir/conjure/java/dialogue/serde/ExceptionDeserializingErrorDecoder.java

```diff
@@ -0,0 +1,293 @@
+            }
+            case 503 -> {
+                return Optional.of(QosException.unavailable(qosReason(response)));
+            }
+        }
+        return Optional.empty();
+    }
+
+    private RuntimeException decodeInternal(Response response) {
+        // Check if the response code indicates an error, and if so, handle it accordingly.
+        Optional<RuntimeException> maybeException = checkCode(response);
+        if (maybeException.isPresent()) {
+            return maybeException.get();
+        }
+        int code = response.code();
+        byte[] body;
+        try (InputStream bodyStream = response.body()) {
+            body = bodyStream.readAllBytes();
+        } catch (NullPointerException | IOException e) {
+            UnknownRemoteException exception = new UnknownRemoteException(code, "<unparseable>");
+            exception.initCause(e);
+            return exception;
+        }
+
+        Optional<String> contentType = response.getFirstHeader(HttpHeaders.CONTENT_TYPE);
+        if (contentType.isEmpty() || !Encodings.matchesContentType(JSON_ENCODING.getContentType(), contentType.get())) {
+            return new UnknownRemoteException(code, new String(body, StandardCharsets.UTF_8));
+        }
+
+        try {
+            String errorName = extractErrorName(body);
+            if (errorName == null) {
+                return createRemoteException(body, code);
+            }
+            DeserializerExceptionPair<?> deserializerExceptionPair = errorNameToExceptionDeserializerMap.get(errorName);
+            if (deserializerExceptionPair == null) {
+                return createRemoteException(body, code);
+            }
+            // Attempt to deserialize the error using the deserializer for the specific exception type.
+            try {
+                AbstractSerializableError<?> error =
+                        deserializerExceptionPair.deserializer().deserialize(new ByteArrayInputStream(body));
+                Type exceptionType = deserializerExceptionPair.exceptionType().getType();
+                @SuppressWarnings("unchecked")
+                Class<? extends RemoteException> exceptionClass =
+                        (Class<? extends RemoteException>) Class.forName(exceptionType.getTypeName());
+                Constructor<? extends RemoteException> exceptionConstructor =
+                        exceptionClass.getConstructor(error.getClass(), int.class);
+                return exceptionConstructor.newInstance(error, code);
+            } catch (Exception e) {
+                // If we're unable to deserialize the error as JSON, throw a RemoteException.
+                log.error("Failed to deserialize error response as exception", SafeArg.of("errorName", errorName), e);
+                return createRemoteException(body, code);
+            }
+        } catch (RemoteException remoteException) {
+            // rethrow the created remote exception
+            return remoteException;
```

Comment:
Out of curiosity, when do we expect this to happen? Nothing in the try block should _throw_ a RemoveException, should it?

Reply:
Removed.